_delphi_utils_python/delphi_utils/smooth.py

```diff
@@ -416,17 +415,24 @@ def savgol_impute(self, signal):
         for ix in np.where(np.isnan(signal))[0]:
             # Boundary cases
             if ix < self.window_length:
-                # A nan following a single value should just be extended
+                # At the boundary, a single value should just be extended
                 if ix == 1:
-                    signal_imputed[ix] = signal_imputed[0]
-                # Otherwise, use savgol fitting
+                    signal_imputed[ix] = signal_imputed[ix - 1]
+                # Reduce the polynomial degree if needed
+                elif ix == 2:
+                    signal_imputed[ix] = self.savgol_predict(
+                        signal_imputed[:ix], 1, -1
+                    )
+                # Otherwise, use savgol fitting on the largest window prior
                 else:
-                    coeffs = self.savgol_coeffs(-ix, -1)
-                    signal_imputed[ix] = signal_imputed[:ix] @ coeffs
-            # Use a polynomial fit on the past window length to impute
+                    signal_imputed[ix] = self.savgol_predict(
+                        signal_imputed[:ix], 2, -1
+                    )
+            # Away from the boundary, use savgol fitting on a fixed window
             else:
-                coeffs = self.savgol_coeffs(-self.window_length, -1)
-                signal_imputed[ix] = (
-                    signal_imputed[ix - self.window_length : ix] @ coeffs
+                signal_imputed[ix] = self.savgol_predict(
+                    signal_imputed[ix - self.window_length : ix],
+                    2,
```

Comment:
How big of a difference is it using a quadratic polynomial vs the previous window_length degree polynomial (assuming im reading this right)?

Reply:
Hm, now I think I should make a different PR and start a discussion there for this. This change is unrelated to the matrix bug.

Reply:
Took this change out.

Reply:
:+1:. the rest of the PR looks good. One thing that I like to do when fixing a bug is write out the test case that would cause the bug and code until it passes. Then just leave it in as a test case and you have some confidence that it's resolved.

Reply:
Nice idea, I'll try that.

Reply:
sounds good, will approve once a test is added

Reply:
Added a test!

Reply:
And a few other bad input tests for the sake of code coverage (couldn't help myself :P).